### ISSUE TITLE
Update filters in types as new implications are added?

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#3642]
In PR #3641, @ThomasBreuer pointed out that installing implications for e.g. "empty lists" does not work, which made me think...

Specifically, I think that it should actually be possible to do so, and the fact that it currently doesn't IMHO seems to hint at a fundamental issue here, closely related to the [method reordering issue #2513](https://github.com/gap-system/gap/issues/2513): Namely, when creating a type object such as `TYPE_LIST_EMPTY_MUTABLE`, we store a set of filters in there exactly once. For this, `NewType` uses `WITH_IMPS_FLAGS` to replace the given filters by their full implications. But then these filters are never updated as new implications are added. Thus, the filters in the type defined in a package potentially depend on the order in which packages are loaded. Granted, usually this doesn't matter, because when using the filters in the type, in many cases `WITH_IMPS_FLAGS` is called again. But not always, as the code added in PR #3641 describes

I thus wonder if we should introduce a major changes, similar to the method reordering one, where for each new implication, we not only update the order of methods, but also update the filters in types. This of course introduce a performance overhead for `InstallTrueMethod`, but we already have put mechanisms in places to take care of that.

Of course the implementation needs some thoughts: How would we even "find" all type objects? With GASMAN, we have code that does that (see the kernel function `COMPACT_TYPE_IDS`) but it doesn't work with other GCs. Another way would be to introduce a "world age" counter, which is incremented each time a new implication is installed. Then, when creating a type, store the current value of the "world age counter" in the type. Then, in the function which queries the flags, add in code like this:
```gap
BIND_GLOBAL( "FlagsType", function(K)
  local flags;
  flags :=  K![POS_FLAGS_TYPE]; 
  if K![POS_CREATIONAGE_TYPE] < WORLD_AGE then
    flags := WITH_IMPS_FLAGS( flags );
    K![POS_FLAGS_TYPE] := flags;
  fi;
  return flags;
end );
```

Thoughts? 

Ping @ChrisJefferson @stevelinton @frankluebeck

**Comment by ChrisJefferson** [comment]
I'd be happy to see it benchmarked, but I'd be worried about slowing down a common route with looking up a global variable, as even at the C level looking up global state has gotten quite expensive nowadays.

Another option, rather than search for type objects in GASMAN, would be to put them all in a weak ptr list. That would obviously have some memory overhead, but I don't think it would be that great.

**Comment by fingolfin** [comment]
In a fresh GAP session with `-A`, there are ~370 type objects (according to `NEW_TYPE_NEXT_ID+2^60`); after `LoadAllPackages();`, I see ~6500. So a weak pointer list would take up about 50kb on a 64bit machine, which indeed should be fine.

**Comment by stevelinton** [comment]
I am a bit wary of automatically updating every type object in the workspace. There may be valid reasons to keep an old (now inconsistent) type lying around and I'm not sure whether its best to change the type of existing objects with a stored type when a new implication is installed.

I think the main need is to catch the types in global variables like `TYPE_LIST_EMPTY_MUTABLE` which are stored ready to be given to new objects. These could be registered when they are created and updated at the same time as method reordering fairly easily.

**Comment by zickgraf** [comment]
For future reference: I just came across two unexpected things in the context of this issue:

1. The documentation of `InstallTrueMethod` claims
> ─────────────────────────────────────────  Example  ────────────────────────────────────────── >    InstallTrueMethod( IsCommutative, IsGroup and IsCyclic );
>  ───────────────────────────────────────────────────────────────────────────────────────────
> 
> Contrary  to  the  immediate  methods  installed with InstallImmediateMethod (78.7-1), logical
> implications  cannot  be  switched  off.  This means that after the above implication has been
> installed,  one can rely on the fact that every object in the filter IsGroup and IsCyclic will
> also be in the filter IsCommutative (35.4-9).

This statement is not true in general due to this issue, as can be seen by the following code:
```
gap> filter := NewFilter( "filter" );;
gap> type := NewType( NewFamily( "family" ), IsGroup and IsCyclic );;
gap> obj := ObjectifyWithAttributes( rec( ), type );;
gap> InstallTrueMethod( filter, IsGroup and IsCyclic );
gap> (IsGroup and IsCyclic)( obj );
true
gap> filter( obj );
false
```

2. When using `ObjectifyWithAttributes`, sometimes the new implications are respected again. This is due to a new type including the new implications being created in
https://github.com/gap-system/gap/blob/3d47e2bc40869ea1b232a0a658d47b1897880fec/lib/type1.g#L798-L799

**Comment by ThomasBreuer** [comment]
I do not catch the point.
This issue just discusses whether we should try to achieve that implications affect also types which were created before the implication got installed.
Thus I do not see why it is surprising that up to now GAP does not apply new implications to old types but applies them to new types.
(If we decide to keep the current behaviour not to update existing types w.r.t. new implications then the documentation of `InstallTrueMethod` should be corrected in the sense that only types created after the installation of an implication will be affected by this implication.)

**Comment by zickgraf** [comment]
Sorry, my wording was not very precise :/ Let me try again:

1. The documentation of `InstallTrueMethod` does not describe the current behaviour. This is unexpected when NOT knowing about this issue.
2. The behaviours of `Objectify` and `ObjectifyWithAttributes` differ. This is unexpected even when knowing about this issue.
3. I just noticed another potential problem: Does updating existing types with new implications also affect existing objects? If not, then the documentation of `InstallTrueMethod` would still not be correct even if existing types would be updated.

Describing the current behaviour correctly in `InstallTrueMethod` seems complicated. Thus I did not create a PR but still wanted to mention the points here so they do not get lost. I hope this makes things clearer :-)

**Comment by ThomasBreuer** [comment]
@zickgraf Thanks.
According to your list:
1. Yes, either the behaviour of `InstallTrueMethod` must be adjusted to the documentation, or the other way round.
2. `Objectify` just sets a gives type (which may not reflect a new implication) whereas `ObjectifyWithAttributes` very likely must create a new type because of the attribute tester filters that shall be set. This issue is about existing types vs. newly created types.
3. Objects point to their types. Thus the problem disappears as soon as the existing types (i.e., type objects) get updated.
